### PR TITLE
issue-2764: add new operator onBackpressureDrop(Action1 onDrop)

### PR DESCRIPTION
**Pull request by stevenzwu**

### CODE REVIEW

File: src/main/java/rx/Observable.java [hunk 1]
```diff
@@ -5270,6 +5270,27 @@ public final Boolean call(T t) {
      *  <dt><b>Scheduler:</b></dt>
      *  <dd>{@code onBackpressureDrop} does not operate by default on a particular {@link Scheduler}.</dd>
      * </dl>
+     *
+     * @param onDrop the action to invoke for each item dropped. onDrop action should be fast and should never block.
+     * @return the source Observable modified to drop {@code onNext} notifications on overflow
+     * @see <a href="http://reactivex.io/documentation/operators/backpressure.html">ReactiveX operators documentation: backpressure operators</a>
+     */
+    public final Observable<T> onBackpressureDrop(Action1 onDrop) {
```
Comment (akarnokd) [inline src/main/java/rx/Observable.java]: Action1 should be generic: ```Action1<? super T> onDrop```

File: src/test/java/rx/BackpressureTests.java [hunk 2]
```diff
@@ -424,25 +428,69 @@ public void testOnBackpressureDrop() {
...
+            assertTrue(NUM - 1 <= lastEvent.intValue());
+            assertTrue(0 < dropCount.get());
+            assertEquals(emitCount.get(), passCount.get() + dropCount.get());
+        }
+    }
+
     @Test(timeout = 10000)
     public void testOnBackpressureDropSynchronous() {
         for (int i = 0; i < 100; i++) {
+            final AtomicInteger dropCount = new AtomicInteger();
```
Comment (akarnokd) [inline src/test/java/rx/BackpressureTests.java]: These changes should be in a separate test method. Leave the ```testOnBackpressureDropSynchronous``` as is and introduce ```testOnBackpressureDropWithActionSynchronous```.

File: src/main/java/rx/Observable.java (hunk 1 above)
Comment (stevenzwu) [inline src/main/java/rx/Observable.java]: will fix. forgot the generic  for arg. did it for member var.

File: src/test/java/rx/BackpressureTests.java (hunk 2 above)
Comment (stevenzwu) [inline src/test/java/rx/BackpressureTests.java]: will do

File: src/main/java/rx/Observable.java [hunk 3]
```diff
@@ -5270,6 +5270,27 @@ public final Boolean call(T t) {
      *  <dt><b>Scheduler:</b></dt>
      *  <dd>{@code onBackpressureDrop} does not operate by default on a particular {@link Scheduler}.</dd>
      * </dl>
+     *
+     * @param onDrop the action to invoke for each item dropped. onDrop action should be fast and should never block.
+     * @return the source Observable modified to drop {@code onNext} notifications on overflow
+     * @see <a href="http://reactivex.io/documentation/operators/backpressure.html">ReactiveX operators documentation: backpressure operators</a>
+     */
+    public final Observable<T> onBackpressureDrop(Action1<T> onDrop) {
```
Comment (akarnokd) [inline src/main/java/rx/Observable.java]: You need ```Action1<? super T>``` here.

File: src/main/java/rx/internal/operators/OperatorOnBackpressureDrop.java [hunk 4]
```diff
@@ -34,7 +37,17 @@
...
     }
-    private OperatorOnBackpressureDrop() { }
+
+    private final Action1<T> onDrop;
+
+    private OperatorOnBackpressureDrop() {
+        this(null);
+    }
+
+    public OperatorOnBackpressureDrop(Action1<T> onDrop) {
```
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorOnBackpressureDrop.java]: You need ```Action1<? super T>``` here.

File: src/main/java/rx/Observable.java (hunk 3 above)
Comment (stevenzwu) [inline src/main/java/rx/Observable.java]: will update. thanks

**Comment by stevenzwu** [comment]
added super to generic type T

File: src/main/java/rx/Observable.java [hunk 5]
```diff
@@ -5270,6 +5270,27 @@ public final Boolean call(T t) {
      *  <dt><b>Scheduler:</b></dt>
      *  <dd>{@code onBackpressureDrop} does not operate by default on a particular {@link Scheduler}.</dd>
      * </dl>
+     *
+     * @param onDrop the action to invoke for each item dropped. onDrop action should be fast and should never block.
+     * @return the source Observable modified to drop {@code onNext} notifications on overflow
+     * @see <a href="http://reactivex.io/documentation/operators/backpressure.html">ReactiveX operators documentation: backpressure operators</a>
+     */
+    public final Observable<T> onBackpressureDrop(Action1<? super T> onDrop) {
```
Comment (akarnokd) [inline src/main/java/rx/Observable.java]: Please add ```@Experimental``` annotation to this method.

**Comment by benjchristensen** [comment]
I will manually add Experimental to it after merging.

**Comment by benjchristensen** [comment]
Thanks @stevenzwu

**Comment by stevenzwu** [comment]
thanks. looks like I missed a lot of details :)

**Pull request merged**